micro-rdk/src/esp32/certificate.rs

```diff
@@ -28,3 +49,277 @@ impl Certificate for WebRtcCertificate {
+            _ => Err(code.into()),
+        }
+    }
+}
+
+impl From<i32> for MbedTLSError {
+    fn from(value: i32) -> Self {
+        MbedTLSError(value)
+    }
+}
+
+impl Display for MbedTLSError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut code = self.0.abs();
+
+        write!(f, "MBEDTLS ERROR ({:#04X}) ", code)?;
+        if (code & 0xFF80) > 0 {
+            let high_level_description = unsafe { mbedtls_high_level_strerr(code) };
+            if !high_level_description.is_null() {
+                let err_str = unsafe { CStr::from_ptr(high_level_description) }
+                    .to_str()
+                    .unwrap();
+                write!(f, "HE : {} ", err_str)?;
+            }
+        }
+        code &= !0xFF80;
+        if code > 0 {
+            let low_level_description = unsafe { mbedtls_low_level_strerr(code) };
+            if !low_level_description.is_null() {
+                let err_str = unsafe { CStr::from_ptr(low_level_description) }
+                    .to_str()
+                    .unwrap();
+                write!(f, "LE : {} ", err_str)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Debug for MbedTLSError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self)
+    }
+}
+
+// TODO consider moving this function in it's own rust wrapper to be reused elsewhere
+// or patch esp-idf-sys upstream to expose it via the bindings
+extern "C" {
+    fn esp_sha(sha_tyep: esp_idf_svc::sys::SHA_TYPE, input: *const u8, len: usize, output: *mut u8);
+}
+
+pub struct GeneratedWebRtcCertificateBuilder {
+    oid_cn: String,
+    oid_c: String,
+    oid_o: String,
+    not_before: String,
+    not_after: String,
+    drbg_context: mbedtls_ctr_drbg_context,
+    entropy: mbedtls_entropy_context,
+    kp_context: mbedtls_pk_context,
+    crt_context: mbedtls_x509write_cert,
+}
+
+impl GeneratedWebRtcCertificateBuilder {
+    pub fn with_common_name(&mut self, cn: String) -> &mut Self {
+        self.oid_cn = cn;
+        self
+    }
+    pub fn with_country(&mut self, c: String) -> &mut Self {
+        self.oid_c = c;
+        self
+    }
+    pub fn with_organization(&mut self, org: String) -> &mut Self {
+        self.oid_o = org;
+        self
+    }
+    pub fn with_notbefore(&mut self, not_before: String) -> &mut Self {
+        self.not_before = not_before;
+        self
+    }
+    pub fn with_notafter(&mut self, not_after: String) -> &mut Self {
+        self.not_after = not_after;
+        self
+    }
+    pub fn build(mut self) -> Result<WebRtcCertificate, MbedTLSError> {
+        unsafe {
+            MbedTLSError::to_unit_result(mbedtls_ctr_drbg_seed(
+                &mut self.drbg_context as *mut mbedtls_ctr_drbg_context,
+                Some(mbedtls_entropy_func),
+                &mut self.entropy as *mut _ as *mut c_void,
+                std::ptr::null(),
+                0,
```

Comment:
- Is `mbedtls_entropy_func` wired up to hardware rng on the esp32 automatically for us in some way? Or should we be looking to connect to it with `entropy_add_source()` and `esp[_fill]_random`?
- Do we want a static personalization string here just for good measure?

Reply:
This is the pattern they use on esp-idf so I am tempted to not stray too much. I am trying to understand whether this is enough for certificate generation.